File: packages/sdk/src/types/events.ts
import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';

export interface AgentRegisteredEvent {
  agent: PublicKey;
  owner: PublicKey;
  name: string;
  priceLamports: BN;
}

export interface AgentUpdatedEvent {
  agent: PublicKey;
  owner: PublicKey;
  name: string;
}

export interface JobCreatedEvent {
  job: PublicKey;
  client: PublicKey;
  agent: PublicKey;
  escrowAmount: BN;
  tokenMint: PublicKey | null;
  autoReleaseAt: BN | null;
}

export interface JobCompletedEvent {
  job: PublicKey;
  agent: PublicKey;
  resultUri: string;
}

export interface JobCancelledEvent {
  job: PublicKey;
  client: PublicKey;
  refundAmount: BN;
}

export interface JobRejectedEvent {
  job: PublicKey;
  agent: PublicKey;
  refundAmount: BN;
}

export interface JobDelegatedEvent {
  parentJob: PublicKey;
  childJob: PublicKey;
  delegatingAgent: PublicKey;
  subAgent: PublicKey;
  amount: BN;
}

export interface PaymentReleasedEvent {
  job: PublicKey;
  agent: PublicKey;
  amount: BN;
  autoReleased: boolean;
  tokenMint: PublicKey | null;
}

export interface DisputeRaisedEvent {
  job: PublicKey;
  raisedBy: PublicKey;
}

export interface DisputeResolvedEvent {
  job: PublicKey;
  refundAmount: BN;
  resolvedBy: PublicKey;
}

export interface AgentRatedEvent {
  agent: PublicKey;
  rater: PublicKey;
  score: number;
  newAvgX100: BN;
}

export interface AgentStakedEvent {
  agent: PublicKey;
  amount: BN;
  totalStake: BN;
}

export interface AgentUnstakedEvent {
  agent: PublicKey;
  amount: BN;
  remainingStake: BN;
}

export interface StakeSlashedEvent {
  agent: PublicKey;
  job: PublicKey;
  slashAmount: BN;
  remainingStake: BN;
}

export interface ArbiterPaidEvent {
  arbiter: PublicKey;
  job: PublicKey;
  feeAmount: BN;
}

export interface AgentProtocolEventMap {
  AgentRegistered: AgentRegisteredEvent;
  AgentUpdated: AgentUpdatedEvent;
  JobCreated: JobCreatedEvent;
  JobCompleted: JobCompletedEvent;
  JobCancelled: JobCancelledEvent;
  JobRejected: JobRejectedEvent;
  JobDelegated: JobDelegatedEvent;
  PaymentReleased: PaymentReleasedEvent;
  DisputeRaised: DisputeRaisedEvent;
  DisputeResolved: DisputeResolvedEvent;
  AgentRated: AgentRatedEvent;
  AgentStaked: AgentStakedEvent;
  AgentUnstaked: AgentUnstakedEvent;
  StakeSlashed: StakeSlashedEvent;
  ArbiterPaid: ArbiterPaidEvent;
}

export type AgentProtocolEventName = keyof AgentProtocolEventMap;

File: agent-protocol/programs/agent-protocol/src/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct AgentRegistered {
    pub agent: Pubkey,
    pub owner: Pubkey,
    pub name: String,
    pub price_lamports: u64,
}

#[event]
pub struct AgentUpdated {
    pub agent: Pubkey,
    pub owner: Pubkey,
    pub name: String,
}

#[event]
pub struct JobCreated {
    pub job: Pubkey,
    pub client: Pubkey,
    pub agent: Pubkey,
    pub escrow_amount: u64,
    pub token_mint: Option<Pubkey>,
    pub auto_release_at: Option<i64>,
}

#[event]
pub struct JobCompleted {
    pub job: Pubkey,
    pub agent: Pubkey,
    pub result_uri: String,
}

#[event]
pub struct JobCancelled {
    pub job: Pubkey,
    pub client: Pubkey,
    pub refund_amount: u64,
}

#[event]
pub struct JobDelegated {
    pub parent_job: Pubkey,
    pub child_job: Pubkey,
    pub delegating_agent: Pubkey,
    pub sub_agent: Pubkey,
    pub amount: u64,
}

#[event]
pub struct PaymentReleased {
    pub job: Pubkey,
    pub agent: Pubkey,
    pub amount: u64,
    pub auto_released: bool,
    pub token_mint: Option<Pubkey>,
}

#[event]
pub struct DisputeRaised {
    pub job: Pubkey,
    pub raised_by: Pubkey,
}

#[event]
pub struct DisputeResolved {
    pub job: Pubkey,
    pub refund_amount: u64,
    pub resolved_by: Pubkey,
}

#[event]
pub struct AgentRated {
    pub agent: Pubkey,
    pub rater: Pubkey,
    pub score: u8,
    pub new_avg_x100: u64,
}

#[event]
pub struct AgentStaked {
    pub agent: Pubkey,
    pub amount: u64,
    pub total_stake: u64,
}

#[event]
pub struct AgentUnstaked {
    pub agent: Pubkey,
    pub amount: u64,
    pub remaining_stake: u64,
}

#[event]
pub struct StakeSlashed {
    pub agent: Pubkey,
    pub job: Pubkey,
    pub slash_amount: u64,
    pub remaining_stake: u64,
}

#[event]
pub struct ArbiterPaid {
    pub arbiter: Pubkey,
    pub job: Pubkey,
    pub fee_amount: u64,
}

#[event]
pub struct JobRejected {
    pub job: Pubkey,
    pub agent: Pubkey,
    pub refund_amount: u64,
}
